src/equations/serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,960 @@
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, NaN)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+function rhs_sgn_upwind!(dq, q, equations, source_terms, cache,
+                         ::Union{BathymetryMildSlope, BathymetryVariable})
+    # Unpack physical parameters and SBP operator `D` as well as the
+    # SBP operator in sparse matrix form `Dmat`
+    g = gravity_constant(equations)
+    (; Dmat_minus) = cache
+    D_upwind = cache.D
+    D = D_upwind.central
+
+    # `q` and `dq` are `ArrayPartition`s. They collect the individual
+    # arrays for the water height `h` and the velocity `v`.
+    eta, v, b = q.x
+    dh, dv, db = dq.x
+    fill!(db, zero(eltype(db)))
+
+    @trixi_timeit timer() "hyperbolic terms" begin
+        # Compute all derivatives required below
+        (; h, h_x, v_x, v_x_upwind, h_hpb_x, b, b_x, hv_x, v2_x,
+        h2_v_vx_x, h_vx_x, p_h, p_0, p_x, tmp,
+        M_h_p_h_bx2, M_h3_3, M_h2_bx) = cache
+        if equations.bathymetry isa BathymetryVariable
+            (; ψ) = cache
+        end
+
+        @. b = equations.eta0 - q.x[3]
+        @. h = eta - b
+        mul!(b_x, D, b)
+
+        mul!(h_x, D, h)
+        mul!(v_x, D, v)
+        mul!(v_x_upwind, D_upwind.minus, v)
+        @. tmp = h * (h + b)
+        mul!(h_hpb_x, D, tmp)
+        @. tmp = h * v
+        mul!(hv_x, D, tmp)
+        @. tmp = v^2
+        mul!(v2_x, D, tmp)
+
+        @. tmp = h^2 * v * v_x
+        mul!(h2_v_vx_x, D, tmp)
+        @. tmp = h * v_x
+        mul!(h_vx_x, D, tmp)
+        # p_0
+        minv6 = -1 / 6
+        @. p_h = minv6 * (h2_v_vx_x
+                          +
+                          h * v * h_vx_x)
+        @. tmp = h * b_x * v^2
+        mul!(p_x, D, tmp)
+        @. p_h += 0.25 * p_x
+        if equations.bathymetry isa BathymetryVariable
+            @. ψ = 0.125 * p_x
+        end
+        @. tmp = b_x * v
+        mul!(p_x, D, tmp)
+        @. p_h += 0.25 * h * v * p_x
+        if equations.bathymetry isa BathymetryVariable
+            @. ψ += 0.125 * h * v * p_x
+        end
+        @. p_0 = p_h * h
+        mul!(p_x, D, p_0)
+        # p_+
+        @. tmp = (0.5 * h * (h * v_x + h_x * v) * v_x_upwind
+                  -
+                  0.25 * (h_x * v + h * v_x) * b_x * v)
+        if equations.bathymetry isa BathymetryVariable
+            @. ψ = 0.125 * (h_x * v + h * v_x) * b_x * v
```

Comment:
```suggestion
            @. psi = 0.125 * (h_x * v + h * v_x) * b_x * v
```
Another question: Do I miss something or do you set $\psi$ already a few lines above, but never use it and here you set it again without using the $\psi$ from before?